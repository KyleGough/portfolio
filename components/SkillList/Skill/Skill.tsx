import { clsx } from 'clsx';
import Image from 'next/image';
import React, { useEffect, useState } from 'react';

import { useObserveElement } from '../../../hooks/useObserveElement';

interface SkillProps {
  type: string;
  progress: number;
  progressClass: string;
  name: string;
  description: string;
  logo: string;
}

export const Skill: React.FC<SkillProps> = ({
  type,
  progress,
  progressClass,
  name,
  description,
  logo,
}) => {
  const [progressRef, isVisible] = useObserveElement<HTMLDivElement>({
    threshold: 0,
  });
  const [animatedIn, setAnimatedIn] = useState(false);
  const durationVar = {
    '--duration-factor': progress,
  } as React.CSSProperties;

  useEffect(() => {
    if (!animatedIn && isVisible) {
      (progressRef.current as HTMLDivElement).classList.replace(
        '-translate-x-full',
        'translate-x-0'
      );
      setAnimatedIn(true);
    }
  }, [progressRef, isVisible, animatedIn]);

  return (
    <div className="flex flex-row justify-center py-6">
      <div className="w-3/12 lg:w-4/12 flex items-center">
        <div className={clsx(type, 'progress overflow-hidden')}>
          <div
            ref={progressRef}
            className={clsx(
              progressClass,
              'custom-duration progress-bar',
              'transition-transform ease-in-out -translate-x-full'
            )}
            style={durationVar}
          ></div>
        </div>
      </div>
      <div className="w-9/12 lg:w-8/12 ml-6 flex items-center">
        <div
          className={clsx(
            { 'opacity-100': animatedIn },
            { 'opacity-0': !animatedIn },
            'mr-4 w-8 h-8 transition-opacity duration-500'
          )}
        >
          <Image src={logo} alt={`${name} logo`} width={30} height={30} />
        </div>
        <div>
          <p>
            <span className="font-extrabold text-lg">{name}</span> -{' '}
            <span className="text-sm">{description}</span>
          </p>
        </div>
      </div>
    </div>
  );
};
