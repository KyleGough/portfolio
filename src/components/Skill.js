import React from 'react';

export default function Skill(props) {
    return (
        <div className='flex flex-row justify-center py-4'>
            <div className='w-3/12 lg:w-4/12 pt-2'>
                <div className={`progress ${props.type}`}>
                    <div className={`progress-bar ${props.progress}`} data-skill={props.progress}></div>
                </div>
            </div>
            <div className='w-9/12 lg:w-8/12 ml-6'>
                <p><span className='font-extrabold'>{props.title}</span> - <span className='text-sm'>{props.description}</span></p>
                <p className='text-sm text-link-hover mt-1'>{props.comment}</p>
            </div>
        </div>
    );
}