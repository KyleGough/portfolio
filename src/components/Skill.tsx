import React, { useState, useEffect, useRef } from 'react';
import useOnScreen from '../hooks/useOnScreen';

type SkillProps = {
    type: string,
    progress: number,
    progressClass: string,
    title: string,
    description: string,
    comment: string,
    logo?: string
};

function Skill(props: SkillProps) {
    const progressRef = useRef<HTMLDivElement>(null);
    const isVisible = useOnScreen(progressRef);
    const [animatedIn, setAnimatedIn] = useState(false);
    const durationVar = { "--duration-factor": props.progress } as React.CSSProperties;

    useEffect(() => {
        if (!animatedIn && isVisible) {
            (progressRef.current as HTMLDivElement).classList.replace('-translate-x-full', 'translate-x-0');
            setAnimatedIn(true);
        }
    }, [isVisible, animatedIn]);

    return (
        <div className='flex flex-row justify-center py-4'>
            <div className='w-3/12 lg:w-4/12 pt-2'>
                <div className={`progress overflow-hidden ${props.type}`}>
                    <div ref={progressRef} className={`custom-duration progress-bar transition-transform ease-in-out -translate-x-full ${props.progressClass}`} style={durationVar}></div>
                </div>
            </div>
            <div className='w-9/12 lg:w-8/12 ml-6 flex'>
                <img loading='lazy' className='w-8 h-8 mr-4' src={props.logo} alt={`${props.title} logo`} />
                <div>
                    <p><span className='font-extrabold'>{props.title}</span> - <span className='text-sm'>{props.description}</span></p>
                    <p className='text-sm text-link-hover mt-1'>{props.comment}</p>
                </div>
            </div>
        </div>
    );
}

export default function renderSkills(skills: [string, number, string, string, string, string][], type: string) {
    return skills.map(skill => {
        return (
            <Skill
                key={skill[0]}
                type={type}
                title={skill[0]}
                progress={skill[1]}
                progressClass={skill[2]}
                description={skill[3]}
                comment={skill[4]}
                logo={skill[5]}
            />
        )
    });
}