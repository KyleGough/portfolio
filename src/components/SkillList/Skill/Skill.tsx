import { clsx } from 'clsx';
import React, { useEffect, useState } from 'react';

import { useObserveElement } from '../../../hooks/useObserveElement';
import { Confidence } from '../SkillList.types';

interface SkillProps {
  type: string;
  progress: number;
  progressClass: string;
  name: string;
  description: string;
  confidence: Confidence;
  logo: string;
}

export const Skill: React.FC<SkillProps> = ({
  type,
  progress,
  progressClass,
  name,
  description,
  confidence,
  logo,
}) => {
  const [progressRef, isVisible] = useObserveElement<HTMLDivElement>({
    threshold: 0,
  });
  const [animatedIn, setAnimatedIn] = useState(false);
  const durationVar = {
    '--duration-factor': progress,
  } as React.CSSProperties;

  useEffect(() => {
    if (!animatedIn && isVisible) {
      (progressRef.current as HTMLDivElement).classList.replace(
        '-translate-x-full',
        'translate-x-0'
      );
      setAnimatedIn(true);
    }
  }, [progressRef, isVisible, animatedIn]);

  return (
    <div className="flex flex-row justify-center py-4">
      <div className="w-3/12 lg:w-4/12 pt-2">
        <div className={clsx(type, 'progress overflow-hidden')}>
          <div
            ref={progressRef}
            className={clsx(
              progressClass,
              'custom-duration progress-bar',
              'transition-transform ease-in-out -translate-x-full'
            )}
            style={durationVar}
          ></div>
        </div>
      </div>
      <div className="w-9/12 lg:w-8/12 ml-6 flex">
        <img
          loading="lazy"
          className={clsx(
            { 'opacity-100': animatedIn },
            { 'opacity-0': !animatedIn },
            'w-8 h-8 mr-4',
            'transition-opacity duration-500'
          )}
          src={logo}
          alt={`${name} logo`}
        />
        <div>
          <p>
            <span className="font-extrabold">{name}</span> -{' '}
            <span className="text-sm">{description}</span>
          </p>
          <p className="text-sm text-link-hover mt-1">{confidence}</p>
        </div>
      </div>
    </div>
  );
};
