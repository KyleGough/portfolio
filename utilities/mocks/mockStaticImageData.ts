const mockStaticImageData = {
  src: '/#image',
  width: 100,
  height: 100,
  blurDataURL: '/#blur',
};

export default mockStaticImageData;
