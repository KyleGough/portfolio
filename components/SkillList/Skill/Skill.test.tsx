import { render, screen } from '@testing-library/react';
import React from 'react';

import { mockIntersectionObserver } from '../../../mocks/mockIntersectionObserver';
import { Skill } from './Skill';

const skill = {
  type: 'progress-web',
  progress: 75,
  progressClass: 'w-[75%]',
  name: 'React',
  description: 'Portfolio, LucidLab, Minesweeper',
  logo: '/#logo',
};

describe('Skill component', () => {
  beforeEach(mockIntersectionObserver);

  it('renders', async () => {
    const { container } = render(<Skill {...skill} />);

    expect(screen.getByText('React')).toBeVisible();
    expect(screen.getByText('Portfolio, LucidLab, Minesweeper')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
});
