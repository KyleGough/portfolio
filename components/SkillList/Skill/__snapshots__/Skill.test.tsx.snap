// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill component renders 1`] = `
<div>
  <div
    class="flex flex-row justify-center py-6"
  >
    <div
      class="w-3/12 lg:w-4/12 flex items-center"
    >
      <div
        class="progress-web progress overflow-hidden"
      >
        <div
          class="w-[75%] custom-duration progress-bar transition-transform ease-in-out -translate-x-full"
          style="--duration-factor: 75;"
        />
      </div>
    </div>
    <div
      class="w-9/12 lg:w-8/12 ml-6 flex items-center"
    >
      <div
        class="opacity-0 mr-4 w-8 h-8 transition-opacity duration-500 relative"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="React logo"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
      <div>
        <p>
          <span
            class="font-extrabold text-lg"
          >
            React
          </span>
           -
           
          <span
            class="text-sm"
          >
            Portfolio, LucidLab, Minesweeper
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;
